gtda/homology/_utils.py

```diff
@@ -1,52 +1,64 @@
-        if n_points == 0 and n_points_to_pad == 0:
-            n_points_to_pad = 1
-
-        if n_points_to_pad > 0:
-            padding = ((0, n_points_to_pad), (0, 0))
-            Xd[dim] = np.pad(Xd[dim], padding, 'constant')
-            Xd[dim][-n_points_to_pad:, :] = \
-                [min_values[dim], min_values[dim]]
-
-    # Add dimension as the third elements of each (b, d) tuple
-    Xd = [np.insert(Xd[dim], 2, dim, axis=1)
-          for dim in homology_dimensions]
-
-    Xd = np.vstack(Xd)
-    return Xd
+def _postprocess_diagrams(
+        Xt, format, homology_dimensions, infinity_values
+        ):
+    # NOTE: `homology_dimensions` must be sorted in ascending order
+    def replace_infinity_values(subdiagram):
+        np.nan_to_num(subdiagram, posinf=infinity_values, copy=False)
+        return subdiagram[subdiagram[:, 0] < subdiagram[:, 1]]
 
+    # Replace np.inf with infinity_values and turn into list of dictionaries
+    # whose keys are the dimensions
+    if format in ["ripser", "flagser"]:  # Input is list of list of subdiagrams
+        # In H0, remove one infinite bar placed at the end by ripser
+        slices = {dim: slice(None) if dim else slice(None, -1)
+                  for dim in homology_dimensions}
+        Xt = [{dim: replace_infinity_values(diagram[dim][slices[dim]])
+               for dim in homology_dimensions} for diagram in Xt]
+    elif format == "gudhi":  # Input is list of list of [dim, (birth, death)]
+        # In H0, remove one infinite bar placed at the beginning by GUDHI
+        slices = {dim: slice(None) if dim else slice(1, None)
+                  for dim in homology_dimensions}
+        Xt = [{dim: replace_infinity_values(
+            np.array([pers_info[1] for pers_info in diagram
+                      if pers_info[0] == dim]).reshape(-1, 2)[slices[dim]]
+            )
+            for dim in homology_dimensions} for diagram in Xt]
+    else:
+        raise ValueError(
+            f"Unknown input format {format} for collection of diagrams."
+            )
 
-def _postprocess_diagrams(Xt, homology_dimensions, infinity_values, n_jobs):
-    # Replacing np.inf with infinity_values and turning list of list of
-    # subdiagrams into list of dictionaries whose keys are the dimensions
-    Xt = [{dim: np.nan_to_num(diagram[dim], posinf=infinity_values)
-          for dim in homology_dimensions}
-          for diagram in Xt]
+    # Conversion to array of triples with padding triples
+    start_idx_per_dim = np.cumsum(
+            [0] + [np.max([len(diagram[dim]) for diagram in Xt] + [1])
+                   for dim in homology_dimensions]
+            )
+    min_values = [min([np.min(diagram[dim][:, 0]) if diagram[dim].size
+                       else np.inf for diagram in Xt])
+                  for dim in homology_dimensions]
+    min_values = [min_value if min_value != np.inf else 0
+                  for min_value in min_values]
+    dtype = next(iter(Xt[0].values())).dtype
```

Comment:
Is this to cover the case where all filtration values are integers?

Reply:
It is. But equally well one could argue that if the user passes a mix of int and float images, this is a bad decision as we would get an error. Perhaps I should always impose float even if occasionally it won't be as efficient? On the other hand, I thought images are a scenario in which one would often get all int arrays.... So I am a bit torn here!

Reply:
I just tested and I think you won't get an error, just silent casting of floats into the new array, which is arguably worse as the user is not get what they are expecting. So this should be handled differently: either always `float` or check whether there is at least one `float`, else `int`.

Reply:
I just checked and cubical and ripser at least always give back float arrays anyway, even if the input is an int distance matrix or int image. I'll just put `float` there.

Reply:
Done.